### PR TITLE
71-coarse-fan-ramp-control

### PR DESCRIPTION
**Pull request by danielcomaduran**
Got the coarse and precise fan control working in Demo Fan Scene 1.
To test the changes play, and select the fanPrefab object. 
In the fabPrefab select positioning mode and set  values for the FanGeneration
- PositioninMode.CenterToBase: Moves to an absolute position based on fan
- PositioningMode.CenterToRails: Moves to a relative position based on ramp rails

When using the PositioninMode.CenterToRails, click on Generate fan after each selection so that a new fan appears in the correct position

Note that there are some bugs because the ramp does not have movement limits

### CODE REVIEW

**Comment by daniellarbourque** [comment]
It looks really good!

CentertoRails and CenterToBase work very well!

I noticed the following about setting R1 and R2:
- If you try to set R1 higher than R2, it automatically is clamped to equal R2. For example, I had R1 = 5 and R2 = 3. When I input R1 = 6, it automatically was changed to 5. 
- However, when I clicked Generate Fan with R1 = R2, it does not generate. 
- So it seems that a fan cannot be generated if R1 = R2. Would it make sense to change the clamping of R1 and R2 so that they can't be equal?

Clicking the fan buttons to rotate and elevate the ramp looks good! @gregbci Right now Dan has FanPresenter.cs call rotate and elevate methods from the model, but in the future should FanClicker call those directly? Similar to how ExamplePresenter works for the temporary rotate/elevate buttons.

**Comment by gregbci** [comment]
@danielcomaduran I also took a peek.  Works great!

Some items:

1.  I agree with @daniellarbourque, FanClicker and FanPresenter are bidirectionally coupled.  I think this is because the classes have overlapping responsibilities.  I would just merge FanClicker into FanPresenter.  This way, FanPresenter is responsible for adding the "ramp control" aspect to a fan.

1.  I think the name "Fan Prefab" can be more specific.  It's a fancy ramp control UI element.  Maybe "Ramp Control Fan" or "Ramp Fan" or "Clickable Fan", something like that. 

1.  The BocciaModel contains ramp settings.  I don't see them being used in the FanPresenter.  I'm not sure if we have a task for this?

1.  Finally, I think FanGenerator should just make fans.  I believe we'll need it for the Game Options screen, where we show the fan shape, maybe called a "Fan Widget".  The SPO stuff should be moved out into FanPresenter.  Or perhaps a separate SPO Fan component?

**Comment by danielcomaduran** [comment]
@gregbci 

Some replies to your last comment.

1. Which responsibilities are shared between FanClicker and FanPresenter? My idea was that FanPresenter would be the component that interacts with the model to present any fan, whether that is the SPO enabled one or the one in the UI settings. This is why I made the FanClicker class, so that would take a Fan and make it interactable. Maybe here I should add the SPO components as well? So then FanPresenter is the class that controls the fan, FanGenerator just makes the individual components, and FanClicker would make the fan elements to be able to be "clicked" with the mouse or BCI. Maybe I should change the name to  FanInteractions?

2. Agreed, I change the name of the prefab to RampControlFan.

3. You mean like Elevation Precision and range in BocciaData?

4. I think this refers to point 1 as well.

Thanks!

**Comment by gregbci** [comment]
@danielcomaduran makes sense, here's some quick answers.  I can chat if you're around too, might be faster.

1. I was thinking the other "presenters" handle clicks, so why not do that in FanPresenter too?  That would avoid the two way dependency (makes code hard to follow).  If you prefer separating 'em have FanClicker talk to the model.  That also breaks the two way dependency.  And yep, the SPO stuff should be part of FanClicker then.

1. thanks!

1. yeah (I assumed the fan needs these things... could be wrong 😄 )

1. yep
